### ISSUE TITLE
createTimer expects Joda's AbstractInstant, not compatible with ZonedDateTime

### ISSUE THREAD

**Issue by kgoderis** [eclipse/smarthome#5562]
org.eclipse.smarthome.model.script classes still rely on org.joda.time.base.AbstractInstant for script actions (e.g. createTimer), whereas we have moved elsewhere in the core to java.time.ZonedDateTime

This leads to errors like 

> Type mismatch: cannot convert from ZonedDateTime to AbstractInstant

in .rules and so forth.

**Comment by sjka** [comment]
What is your take on backward compatibility wrt to existing scripts?

**Comment by sjka** [comment]
Nevermind my previous question. I just realized this is only about `ScriptExecution.createTimer(...)` and `ScriptExecution.createTimerWithArgument(...)`. These can be kept compatible by simply by method overloading.
Will you create a PR for it?

**Comment by kgoderis** [comment]
@sjka Whilst giving it a go I started to wonder if we need to keep the backward compatibility with Joda. The API was already broken/depreciated when moving to ZonedDateTime, so why not changing everything else? 

Philosophical question : it is OK to use instanceof while doing generics? In casu, the Time and TimerImpl classes have to altered. This can be done via generics, or via an abstract class and some inheritance. In case of the former, you would end up with code like:

```
    private D startTime;

    public boolean reschedule(D newTime) {
        try {
            Trigger trigger = null;
            if (startTime instanceof ZonedDateTime) {
                trigger = newTrigger().startAt(Date.from(((ZonedDateTime) newTime).toInstant())).build();
            }
            if (startTime instanceof AbstractInstant) {
                trigger = newTrigger().startAt(((AbstractInstant) newTime).toDate()).build();

            }
...
```

in TimerImpl.java. I find that a bit of a hack myself.

**Comment by flaviocosta-net** [comment]
@kgoderis, `instanceof` and type casts are _runtime_ typing mechanisms. One of the biggest advantages of introducing generics in Java was to allow type checking at _compile time_.

If you are writing code with generics, you should try to take advantage of that as much as possible, and the fact you may still need `instanceof` often indicates that generics is not producing the expected benefits. If your scenario does not allow you to avoid it by using generics, you may probably want to implement it in another way - for instance, with method overloading.

**Comment by kaikreuzer** [comment]
>  The API was already broken/depreciated when moving to ZonedDateTime, so why not changing everything else?

I just came across this issue when looking at https://github.com/openhab/openhab-docs/pull/693 - I thus agree with @kgoderis that the DateTime usage within script/rules has been (accidentially) broken with the ZonedDateTime, so we should probably get rid off Jodatime from the rules completely asap. @sjka Do you agree or am I overlooking anything?

**Comment by sjka** [comment]
No, you are right I guess. Let's move forward then and kick joda-time out completely.
